Add the density graph to ScreenSelectMusic

teejusb opened the pull request:
This is a WIP for adding the Density Graph to ScreenSelectMusic. To make it easy to review, I figured I'd start the PR now so people can follow along so it can be incrementally reviewed to make the process easier.

There are still additional changes that should be made to make the experience better for end users. For example, it'll currently lag when trying to parse really long songs. As a part of this PR, I'll be refactoring how all chart parsing mechanisms work to leverage caching to help speed up this process. Some examples include -- hash generation, density graph, breakdown calculation.

The current implementation looks like this:
![image](https://user-images.githubusercontent.com/5017202/111124388-b4e71e00-852d-11eb-8a6d-fbb91ce596b8.png)

And in Widescreen:
![image](https://user-images.githubusercontent.com/5017202/111124453-c5979400-852d-11eb-88d0-8df2d6642da1.png)

teejusb committed:
Initial commit of adding the density graph into ScreenSelectMusic

- Added graph
- Adjusted difficulty picker

TODO:
- Fix up chart parsing code to do more caching to better the user experience

teejusb committed:
Convert DensityGraph.lua to tabs for consistency

teejusb committed:
Disable Density Graph for course mode

Adjust the course song list pane to better center it as a result of moving the banner + song info. The pane can be potentially be expanded later to better make use of the extra space

natano commented on Scripts/SL-Histogram.lua:
```diff
@@ -96,7 +96,13 @@ function NPS_Histogram(player, width, height)
 		InitCommand=function(self)
 			self:SetDrawState({Mode="DrawMode_QuadStrip"})
 		end,
-		CurrentSongChangedMessageCommand=function(self)
```
I'm not sure if CurrentSongChangedMessageCommand is still necessary for course mode.

teejusb replied:
I just checked and it seems like CurrenSongChangedMessageCommand is not necessary for course mode :)

natano commented on BGAnimations/ScreenSelectMusic overlay/PerPlayer/DensityGraph.lua:
```diff
@@ -0,0 +1,137 @@
+local width = IsUsingWideScreen() and 286 or 276
+
+-- Max height at 280 BPM 16ths. Anything faster will just scale the graph. 
+local maxNps = 280/60 * 4
```
Do you intend to eventually scale all the graphs to 280bpm stream height? I think that might be very discouraging for "casual" players. How about scaling the graphs to the max NPS across all charts in the song? That way the graphs would still be comparable across difficulties, but we would avoid situations where all of the stuff I play only fills the graph height about half way.

teejusb replied:
Yeah the original implementation scaled it to 280 bpm stream height, but I realized the experience wasn't great for the same reason you described. Often times the graph was just completely covered by the breakdown text quad so I got rid of it  (so this is actually dead code at the moment that I forgot to remove). I do like the idea of comparing graphs across all difficulties of a song though.  I think I will implement that and the planned refactor will help with that approach.

natano commented on BGAnimations/ScreenSelectMusic overlay/PerPlayer/DensityGraph.lua:
```diff
@@ -0,0 +1,137 @@
+
+-- The Density Graph itself
+af[#af+1] = NPS_Histogram(player, width, height)..{
+	Name="Histogram",
```
I think SM is not very happy about setting the name here:
```
/////////////////////////////////////////
WARNING: #1  ... in /Themes/_fallback/Scripts/02 ActorDef.lua:30
WARNING: #2  (tail call)
WARNING: #3  main in /Themes/Simply-Love-SM5/BGAnimations/ScreenSelectMusic overlay/PerPlayer/DensityGraph.lua:70
WARNING: #4  (tail call)
WARNING: #5  main in /Themes/Simply-Love-SM5/BGAnimations/ScreenSelectMusic overlay/PerPlayer/Over.lua:10
WARNING: #6  (tail call)
WARNING: #7  main in /Themes/Simply-Love-SM5/BGAnimations/ScreenSelectMusic overlay/default.lua:30
WARNING: 
WARNING: Overriding "Name": string with string
/////////////////////////////////////////
```
When scrolling through the songwheel the log gets spammed by those errors.

teejusb replied:
Ah it's probably because the returned actor already has a name. I'll remove the name from here.

teejusb replied:
Actually, I'm going to remove the name from the SL-Histogram.lua as it doesn't seem to be referenced anywhere else.

natano commented on BGAnimations/ScreenSelectMusic overlay/PerPlayer/DensityGraph.lua:
```diff
@@ -0,0 +1,137 @@
+	end,
+
+	CurrentSongChangedMessageCommand=function(self) self:queuecommand("UpdateGraph") end,
+	CurrentCourseChangedMessageCommand=function(self) self:queuecommand("UpdateGraph") end,
```
Maybe not required because of the StepsP*Changed commands? (not tested)

teejusb replied:
Looks like both of those are indeed not required. Removed.

natano commented on BGAnimations/ScreenSelectMusic overlay/PerPlayer/DensityGraph.lua:
```diff
@@ -0,0 +1,137 @@
+
+		if player == PLAYER_2 then
+			self:addy(height+24)
+		end
```
Should we vertically center or scale up the NPS graph when there is only one player?

Pro:
- Use dead space and more symmetry

Contra:
- Consistency. Players get used to always find info in the same place.

teejusb replied:
I don't think we should scale up since the NPS graph looks kinda weird the more "square" it gets. I could vertically center, but I think for now I do prefer the consistency of the placements. It also makes the code a bit simpler :)

natano replied:
fair

natano commented on BGAnimations/ScreenSelectMusic overlay/PerPlayer/over.lua:
```diff
@@ -1,10 +1,20 @@
+		t[#t+1] = LoadActor("./DensityGraph.lua", player)
+	end
+	-- AuthorCredit, Description, and ChartName associated with the current stepchart
+	t[#t+1] = LoadActor("./StepArtist.lua", player)
```
Is there any difference between this being in over or under?

teejusb replied:
I don't think so since there aren't any overlapping elements. I originally put it in over since I was keeping the original elements intact below it as I was working. We could probably just combine the under and over files if we want to. WDYT?

natano replied:
I'd probably put all of it in under.

teejusb replied:
I removed over.lua, moved everything to under.lua, and then renamed under.lua to default.lua

Horsey- commented:
(I haven't downloaded the beta fork yet so this might be moot)

This would be a great time to implement the difficulty selector functionality whereby edit charts show automatically in the list of difficulties. In the default SM5 theme you can see edit difficulties without having to scroll to "find" them. It would eliminate the need to have the "has edit" graphic and be more streamlined as such. You could even reclaim that lost space in the music wheel for other things (make the song wheel narrower to fit more elements onscreen? wider NPS graphs?)

teejusb commented:
For the time being, we are still keeping the edit functionality as is. We decided that the 5 boxes should coincide with the 5 possible difficulty spots for a chart. This might be something we will revisit in the future but it's not currently in the scope of this PR.

teejusb committed:
Couple minor cleanups based off of Natano's review. Removed names of graphs in in SL-Histogram.lua as they're not referenced anyways.

teejusb committed:
Cache and reference values when calling ChartParser

This commit does the following:

 - Rearranges SL-ChartParser.lua to reference the SL.Streams Cache to determine if we need to parse the file or not
- Create a generic GetMeasureInfo that fetches the PeakNPS, NPSperMeasure, and NotesPerMeasure values which get stored in the cache
- Move GetStreamSequences to MeasureCounterAndModsLevel.lua as that's the only place it's called. It uses the NotesPerMeasure calculated earlier to determine to compute the stream sequences

teejusb committed:
Remove an extra Trace

natano commented on BGAnimations/ScreenGameplay in/MeasureCounterAndModsLevel.lua:
```diff
@@ -1,3 +1,67 @@
+	if #streamMeasures > 0 then
+		local breakStart = 0
+		local k, v = next(streamMeasures) -- first element of a table
+		local breakEnd = streamMeasures[k] - 1
```
Isn't `streamMeasures[k]` == `v`? Or am I misunderstanding how this works?

teejusb replied:
You're correct. I don't actually remember my reasoning, but I think it's because below we use `streamMeasure[k+1]` for the `nextVal` so that it looked reasonable to also keep it as `streamMeasure[k]`. I renamed `v` to `curVal` which I think makes thing a bit nicer

natano commented on BGAnimations/ScreenGameplay in/MeasureCounterAndModsLevel.lua:
```diff
@@ -1,3 +1,67 @@
+	local streamSequences = {}
+	-- Count every measure as stream/non-stream.
+	-- We can then later choose how we want to display the information.
+	local measureSequenceThreshold = 1
```
Why is this a variable? Would you ever use anything else but 1 for it?

teejusb replied:
Talked offline. I agree we can use 1 everywhere, but I think this function can actually use a refactor to make it a bit more simpler if we're using that assumption. I added a TODO for now that I'll clean up later.

natano commented on BGAnimations/ScreenSelectMusic overlay/PerPlayer/DensityGraph.lua:
```diff
@@ -0,0 +1,132 @@
+}
+
+-- The Density Graph itself
+-- The name "DensityGraph_AMV" comes from SL-Histogram.lua
```
obsolete comment

teejusb replied:
Removed.

natano commented on BGAnimations/ScreenSelectMusic overlay/PerPlayer/DensityGraph.lua:
```diff
@@ -0,0 +1,132 @@
+		end,
+		CurrentStepsP2ChangedMessageCommand=function(self)
+			self:queuecommand("UpdateBreakdown")
+		end,
```
Can we also do the `["CurrentSteps"..pn.."ChangedMessageCommand"]` trick here?

teejusb replied:
Ooh good catch. Done.

natano commented on BGAnimations/ScreenSelectMusic overlay/PerPlayer/DensityGraph.lua:
```diff
@@ -0,0 +1,132 @@
+	},
+	
+	LoadFont("Miso/_miso")..{
+		Text="No Streams!",
```
For the future: I think it would be nice if no text and background quad would be displayed if we actually didn't find any streams in the chart.

teejusb replied:
Agreed! My plan right now is to display streams if it's a stream chart, display tech stuff if it's a tech chart, or hide this altogether if it's not deemed as either.

natano reviewed (approved):
`GetStreamSequences` was hardest to understand for me. I can't fully rule out that I missed some issue in it. In general i think it could work in one pass instead of creating the streamMeasures table and then iterating over it, but that's not really important.

I'm fine with merging the branch.

teejusb committed:
Clean up PR based off of Natano's review comments.
